# Final Project

**Họ và tên thành viên 1:** Trần Tuấn Đạt

**MSSV:** 18120313

**Họ và tên thành viên 2:** Bùi Trọng Hoàng

**MSSV:** 18120537

---

## Thư viện sử dụng

In [29]:
import pandas as pd
import numpy as np

## Chú thích

1. Phim

Để cho ngắn gọn, thì mỗi dòng của dữ liệu sẽ được gọi là một dòng 'phim'

Điều này để rút gọn giữa Movie hay TV Show

Trong trường hợp cụ thể thì sẽ gọi một dòng là Movie hoặc TV Show (Shơw)

---

# I. Thu thập dữ liệu

Dữ liệu: Netflix Movies and TV Shows

Mô tả dữ liệu:

- Tập hợp dữ liệu tất cả các TV Shows và phim trên nền tảng Netflix. Được cập nhật hàng tháng


Nguồn dữ liệu:

- Link: https://www.kaggle.com/shivamb/netflix-shows

- Nguồn sinh dữ liệu: Flixable

- Phương pháp thu thập: Regular API calls

- Chủ sỡ hữu Dataset: Shivam Bansal

Thông tin phiên bản của bộ dữ liệu: 
- Expected update frequency: Quarterly
- Last updated: 2021-09-27
- Date created: 2019-12-04
- Current version: Version 5


Thông tin về quyền sử dụng:

- License: CC0: Public Domain

    You can copy, modify, distribute and perform the work, 
    even for commercial purposes, all without asking permission.
    
 https://creativecommons.org/publicdomain/zero/1.0/
 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x



In [3]:
df = pd.read_csv('Data/netflix_titles.csv')

---

# II. Khám phá dữ liệu

### 1. Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

#### Số cột 

In [4]:
n_cols = len(df.columns)
n_cols

12

#### Số dòng

In [77]:
n_rows = df.shape[0]
n_rows

8807

### 2. Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không? Dữ liệu có các dòng bị lặp không?

#### Mỗi dòng có ý nghĩa gì?

Mỗi dòng là đại diện cho một bộ phim hoặc một TV show

Chỉ có một dòng đầu tiên là chứa các tên cột, còn không có dòng lạc loài trong phần còn lại của dữ liệu

#### Dữ liệu có các dòng bị lặp không?

In [6]:
# True = 1, False = 0, nên ta sum cột này lại, nếu = 0 thì không có dòng trùng
res = df.duplicated().sum()

#kết quả là không có dòng trùng nhau
res > 0


False

### 3. Mỗi cột có ý nghĩa gì? Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

#### Mỗi cột có ý nghĩa gì?

Theo mô tả bộ dữ liệu tại nguồn

- **'show_id'** : Mã id của phim

- **'type'**: Loại của phim là Movie hay TV Show

- **'title'**: Tên phim

- **'director'**: Đạo diễn của phim

- **'cast'**: Diễn viên của phim

- **'country'**: Nước sản xuất

- **'date_added'**: Ngày được thêm vào bộ dữ liệu

- **'release_year'**: Năm sản xuất của phim

- **'rating'**: Phân loại đối tượng xem của phim (PG-13, TV-MA,..)

- **'duration'**: Thời lượng của phim

- **'listed_in'**: Thể loại của phim

- **'description'**: Mô tả sơ bộ của phim
     

#### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [7]:
df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

Các cột dữ liệu đều có kiểu String, chỉ riêng cột release_year là kiểu số nguyên

Điều này là bình thường không có dữ liệu không phù hợp

Cột date_added có kiểu String, nhưng hiện tại chưa có mục đích xử lý nên không chuyển sang kiểu Datetime

Nếu cần sẽ chuyển sang kiểu Datetime trong phần tiền xử lý cho câu hỏi riêng

### 4. Numerical

#### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

Cột numerical có hai cột *release_year* và cột *date_added*, nên ta sẽ khảo sát hai cột này.

Nhưng hiện tại cột *date_added* đang có kiểu dữ liệu là **string** nên ta cần phải chuyển về dưới dạng **datetime** để tiến hành khảo sát

#### Chuyển cột *date_added* về kiểu datetime

In [42]:
def convertDateTime(date):
    try:
        return pd.to_datetime(date,format="%B %d, %Y")
    except:
        return np.nan
date_add = df['date_added'].apply(convertDateTime)
df['date_added'] = date_add

0      2021-09-25
1      2021-09-24
2      2021-09-24
3      2021-09-24
4      2021-09-24
          ...    
8802   2019-11-20
8803   2019-07-01
8804   2019-11-01
8805   2020-01-11
8806   2019-03-02
Name: date_added, Length: 8807, dtype: datetime64[ns]

In [44]:
# cột numerical là cột release_year và cột date_added ta sẽ khảo sát hai cột này
rls_col = df['release_year']
date_add = df['date_added']

98

#### Số-lượng/tỉ-lệ các giá trị thiếu?

In [45]:
n_year_missing = rls_col.isna().sum()# == 0
ratio_year_missing = n_year_missing * 100 / len(rls_col) #== 0

n_date_missing = date_add.isna().sum()#==98
ratio_date_missing = n_date_missing * 100 / len(date_add)#== 1.11275

Cột *release_year* **không có** giá trị nào bị thiếu, tỉ lệ giá trị bị thiếu là **0%**

Cột *date_added* có **98** giá trị bị thiếu, tỉ lệ giá trị bị thiếu là **1.11275%**

#### Min? max? → Có gì bất thường không?

In [50]:
year_earliest = rls_col.min() #1925
date_earliest = date_add.min()#Timestamp('2008-01-01 00:00:00')

In [52]:
year_latest = rls_col.max() #2021
date_latest = date_add.max()#Timestamp('2021-09-25 00:00:00')

Không có gì bất thường

### 5. Categorical

#### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

In [56]:
categorical_df = df.drop('release_year', axis=1)
categorical_df = categorical_df.drop('date_added', axis = 1)

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country rating  \
0                                                   NaN  United States  PG-13   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa  TV-MA   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN  TV-MA   
3                                                   NaN            NaN  TV-MA   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India  TV-MA   
...                                                 ...            ...    ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States      R   
8803                                                NaN            NaN  TV-Y7   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States      R   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States     PG   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India  TV-14   

       duration                                          listed_in  \
0        90 min                                      Documentaries   
1     2 Seasons    International TV Shows, TV Dramas, TV Mysteries   
2      1 Season  Crime TV Shows, International TV Shows, TV Act...   
3      1 Season                             Docuseries, Reality TV   
4     2 Seasons  International TV Shows, Romantic TV Shows, TV ...   
...         ...                                                ...   
8802    158 min                     Cult Movies, Dramas, Thrillers   
8803  2 Seasons             Kids' TV, Korean TV Shows, TV Comedies   
8804     88 min                            Comedies, Horror Movies   
8805     88 min                 Children & Family Movies, Comedies   
8806    111 min     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 10 columns]

#### Số lượng các giá trị khác nhau? Show một vài giá trị → Có gì bất thường không?

In [57]:
categorical_df.agg([pd.Series.nunique, pd.Series.unique])

show_id              type  \
nunique                                               8807                 2   
unique   [s1, s2, s3, s4, s5, s6, s7, s8, s9, s10, s11,...  [Movie, TV Show]   

                                                     title  \
nunique                                               8807   
unique   [Dick Johnson Is Dead, Blood & Water, Gangland...   

                                                  director  \
nunique                                               4528   
unique   [Kirsten Johnson, nan, Julien Leclercq, Mike F...   

                                                      cast  \
nunique                                               7692   
unique   [nan, Ama Qamata, Khosi Ngema, Gail Mabalane, ...   

                                                   country  \
nunique                                                748   
unique   [United States, South Africa, nan, India, Unit...   

                                                    rating  \
nunique                                                 17   
unique   [PG-13, TV-MA, PG, TV-14, TV-PG, TV-Y, TV-Y7, ...   

                                                  duration  \
nunique                                                220   
unique   [90 min, 2 Seasons, 1 Season, 91 min, 125 min,...   

                                                 listed_in  \
nunique                                                514   
unique   [Documentaries, International TV Shows, TV Dra...   

                                               description  
nunique                                               8775  
unique   [As her father nears the end of his life, film...

Kết luận: không có gì bất thường 

    vì có đúng 8807 giá trị id khác nhau, có 2 loại phim là Movie và TV Show, 
    các cột còn lại không có cột số lượng unique vượt số dòng thực

    tuy nhiên, cột description, date_added bị thiếu giá trị ở một vài dòng
    cột director, cast, country, rating, duration có nhiều giá trị trùng nhau, hoặc bị khuyết

#### Số giá trị bị thiếu và tỉ lệ giá trị bị thiếu

In [75]:
categorical_missing_value = categorical_df.agg([pd.Series.isna], axis=1).sum()
categorical_missing_ratio = categorical_missing_value * 100 / n_rows

In [73]:
categorical_missing_value

show_id           0
type              0
title             0
director       2634
cast            825
country         831
rating            4
duration          3
listed_in         0
description       0
dtype: int64

In [76]:
categorical_missing_ratio

show_id         0.000000
type            0.000000
title           0.000000
director       29.908028
cast            9.367549
country         9.435676
rating          0.045418
duration        0.034064
listed_in       0.000000
description     0.000000
dtype: float64

Những cột có giá trị thiếu là *director, cast, country, rating, duration*, tỉ lệ giá trị bị thiếu đạt cao ở cột *director* với tỉ lệ **29.908028%**, các cột *show_id, type, title, listed_in, description* không có giá trị nào bị mất.

# III. Đưa ra câu hỏi để trả lời

---

### Tiền xử lý

---

### Phân tích dữ liệu để trả lời